File: backend/app/services/rag_service.py
"""RAG — SQLite-based retrieval of high-engagement reference comments"""
import logging
from app.db.database import get_conn

log = logging.getLogger("rag")


def get_examples(post_type: str, topic: str = "", limit: int = 5) -> list[dict]:
    try:
        conn = get_conn()
        rows = conn.execute(
            "SELECT comment_text, post_content, comment_angle, engagement_score, likes "
            "FROM reference_comments WHERE post_type = ? ORDER BY engagement_score DESC LIMIT ?",
            (post_type, limit)
        ).fetchall()

        if len(rows) < 3 and topic:
            keywords = [w for w in topic.split() if len(w) > 3]
            if keywords:
                clause = " OR ".join(["post_content LIKE ?" for _ in keywords])
                params = [f"%{k}%" for k in keywords] + [limit - len(rows)]
                more = conn.execute(
                    f"SELECT comment_text, post_content, comment_angle, engagement_score, likes "
                    f"FROM reference_comments WHERE {clause} ORDER BY engagement_score DESC LIMIT ?",
                    params
                ).fetchall()
                rows = list(rows) + list(more)

        conn.close()
        results = [dict(r) for r in rows]
        log.info(f"RAG: {len(results)} examples for type='{post_type}'")
        return results
    except Exception as e:
        log.error(f"RAG error: {e}")
        return []


def format_for_prompt(examples: list[dict]) -> str:
    if not examples:
        return ""
    lines = ["\n===== HIGH-ENGAGEMENT EXAMPLES =====",
             "Learn from these real comments that got high engagement:\n"]
    for i, ex in enumerate(examples):
        lines.append(f"Example {i+1}: {ex.get('comment_angle', 'GENERAL')}")
        lines.append(f"Post: \"{str(ex.get('post_content', ''))[:150]}...\"")
        lines.append(f"Comment: \"{ex.get('comment_text', '')}\"")
        lines.append(f"Engagement: {ex.get('likes', 0)} likes\n")
    lines.append("===== END EXAMPLES =====")
    return "\n".join(lines)

File: src/lib/services/rag-service.ts
/**
 * RAG Service — SQLite-based retrieval of high-engagement reference comments
 * Replaces ChromaDB with simple SQLite FTS matching
 */
import { getSqlite } from "../db";

export interface RagResult {
  commentText: string;
  postContent: string;
  angle: string;
  engagementScore: number;
  likes: number;
}

/**
 * Retrieve similar high-engagement comments for a post type
 */
export function getRelevantExamples(
  postType: string,
  mainTopic: string,
  limit = 5
): RagResult[] {
  try {
    const sqlite = getSqlite();

    // First try: match by post_type
    let results = sqlite
      .prepare(
        `SELECT comment_text, post_content, comment_angle, engagement_score, likes
         FROM reference_comments
         WHERE post_type = ?
         ORDER BY engagement_score DESC
         LIMIT ?`
      )
      .all(postType, limit) as any[];

    // Fallback: keyword search in post content
    if (results.length < 3 && mainTopic) {
      const keywords = mainTopic.split(" ").filter((w: string) => w.length > 3);
      if (keywords.length > 0) {
        const likeClause = keywords
          .map(() => "post_content LIKE ?")
          .join(" OR ");
        const likeParams = keywords.map((k: string) => `%${k}%`);

        const moreResults = sqlite
          .prepare(
            `SELECT comment_text, post_content, comment_angle, engagement_score, likes
             FROM reference_comments
             WHERE ${likeClause}
             ORDER BY engagement_score DESC
             LIMIT ?`
          )
          .all(...likeParams, limit - results.length) as any[];

        results = [...results, ...moreResults];
      }
    }

    return results.map((r: any) => ({
      commentText: r.comment_text,
      postContent: r.post_content,
      angle: r.comment_angle,
      engagementScore: r.engagement_score,
      likes: r.likes,
    }));
  } catch (error) {
    console.error("[RAG] Error:", error);
    return [];
  }
}

/**
 * Format RAG results for prompt injection
 */
export function formatRagForPrompt(examples: RagResult[]): string {
  if (examples.length === 0) return "";

  let formatted = "\n===== HIGH-ENGAGEMENT EXAMPLES =====\n";
  formatted +=
    "These real comments received high engagement. Learn from their angles and style:\n\n";

  for (let i = 0; i < examples.length; i++) {
    const ex = examples[i];
    formatted += `Example ${i + 1}: ${ex.angle || "GENERAL"}\n`;
    formatted += `Post: "${ex.postContent?.slice(0, 150)}..."\n`;
    formatted += `Comment: "${ex.commentText}"\n`;
    formatted += `Engagement: ${ex.likes} likes (score: ${ex.engagementScore})\n\n`;
  }

  formatted += "===== END EXAMPLES =====\n";
  return formatted;
}
